File: android/src/main/java/com/iologinutils/IoLoginUtilsModule.java
package com.iologinutils;


import androidx.annotation.NonNull;

import com.facebook.react.bridge.Arguments;
import com.facebook.react.bridge.Promise;
import com.facebook.react.bridge.ReactApplicationContext;
import com.facebook.react.bridge.ReactContextBaseJavaModule;
import com.facebook.react.bridge.ReactMethod;
import com.facebook.react.bridge.ReadableMap;
import com.facebook.react.bridge.WritableArray;
import com.facebook.react.module.annotations.ReactModule;

import java.io.IOException;
import java.net.HttpURLConnection;
import java.net.URL;
import java.util.ArrayList;
import java.util.HashMap;

import android.app.Activity;

import android.net.Uri;

import androidx.browser.customtabs.CustomTabsIntent;


@ReactModule(name = IoLoginUtilsModule.NAME)
public class IoLoginUtilsModule extends ReactContextBaseJavaModule {
  public static final String NAME = "IoLoginUtils";

  public IoLoginUtilsModule(ReactApplicationContext reactContext) {
    super(reactContext);
  }

  @Override
  @NonNull
  public String getName() {
    return NAME;
  }


  @ReactMethod
  public static void getRedirects(String url, ReadableMap headers, Promise promise) throws IOException {
        ArrayList<String> urlArray = new ArrayList<String>();
        try {
            findRedirects(url, headers, urlArray);
            String[] urls = urlArray.toArray(new String[0]);
            WritableArray resultArray = Arguments.fromArray(urls);
            promise.resolve(resultArray);
          }
        catch (IOException e) {
          promise.reject("error", e.getMessage());
        }
    }

    public static void findRedirects(String url, ReadableMap headers, ArrayList<String> urlArray) throws IOException {
        HttpURLConnection connection = (HttpURLConnection) new URL(url).openConnection();
        connection.setInstanceFollowRedirects(false);
        HashMap<String, Object> headersMap = headers.toHashMap();
        for (String key : headersMap.keySet()) {
            String value = headersMap.get(key).toString();
            connection.setRequestProperty(key, value);
        }
        int responseCode = connection.getResponseCode();
        if (responseCode >= 300 && responseCode <= 399) {
            String redirectUrl = connection.getHeaderField("Location");
            if(redirectUrl.startsWith("/")){
                URL previousUrl = new URL(url);
                String redirectScheme = previousUrl.getProtocol();
                String redirectHost = previousUrl.getHost();
                String port = String.valueOf(previousUrl.getPort());
                redirectUrl = redirectScheme +"://" + redirectHost + (port.equals("-1") ? "" : ":" + port) + redirectUrl;
            }
            urlArray.add(redirectUrl);
            findRedirects(redirectUrl,urlArray);
        } else {
            return;
        }
    }

    public static void findRedirects(String url, ArrayList<String> urlArray) throws IOException {
        HttpURLConnection connection = (HttpURLConnection) new URL(url).openConnection();
        connection.setInstanceFollowRedirects(false);
        int responseCode = connection.getResponseCode();
        if (responseCode >= 300 && responseCode <= 399) {
            String redirectUrl = connection.getHeaderField("Location");
            if (redirectUrl.startsWith("/")) {
                URL previousUrl = new URL(url);
                String redirectScheme = previousUrl.getProtocol();
                String redirectHost = previousUrl.getHost();
                String port = String.valueOf(previousUrl.getPort());
                redirectUrl = redirectScheme +"://" + redirectHost + (port.equals("-1") ? "" : ":" + port) + redirectUrl;
            }
            urlArray.add(redirectUrl);
            findRedirects(redirectUrl, urlArray);
        } else {
            return;
        }
    }


    @ReactMethod
    public void openAuthenticationSession(String url, String callbackURLScheme, Promise promise) {
      CustomTabActivity.promise = promise;
      prepareCustomTab(Uri.parse(url),promise);
    }

  public synchronized void prepareCustomTab(Uri uri, Promise promise) {

    Activity activity = getCurrentActivity();
      try{
        CustomTabsIntent.Builder intentBuilder = new CustomTabsIntent.Builder();
        CustomTabActivityHelper customTabHelper = new CustomTabActivityHelper();

        CustomTabActivity.customTabContext = activity;
        CustomTabActivity.customTabHelper = customTabHelper;

        customTabHelper.bindCustomTabsService(activity,uri);

        synchronized(CustomTabActivityHelper.lock) {
          while(CustomTabActivityHelper.mClient == null) {
            try {
              CustomTabActivityHelper.lock.wait();
            } catch (InterruptedException e) {
              promise.reject("error", e.getMessage());
            }
          }
        }
        customTabHelper.mayLaunchUrl(uri);
        CustomTabsIntent intent = intentBuilder.build();
        CustomTabActivityHelper.openCustomTab(activity,intent,uri,promise);
      } catch(Exception e){
        promise.reject("error", e.getMessage());
      }

  }


}

File: android/src/main/java/com/iologinutils/IoLoginUtilsModule.kt
package com.iologinutils

import com.facebook.react.bridge.*
import com.facebook.react.module.annotations.ReactModule
import com.iologinutils.IoLoginError.Companion.generateErrorUserInfo
import com.iologinutils.browser.BrowserPackageHelper
import java.io.IOException
import java.net.HttpURLConnection
import java.net.URL

@ReactModule(name = IoLoginUtilsModule.NAME)
class IoLoginUtilsModule(reactContext: ReactApplicationContext?) :
  ReactContextBaseJavaModule(reactContext) {

  //region custom tabs
  @ReactMethod
  fun openAuthenticationSession(
    url: String,
    callbackURLScheme: String,
    @Suppress("UNUSED_PARAMETER") shareiOSCookies: Boolean,
    promise: Promise) {
    authorizationPromise = promise

    currentActivity?.let { ioActivity ->
      AuthorizationService(ioActivity).use { service ->
        service.performAuthorizationRequest((url))
      }
    } ?: run {
      rejectAndClearAuthorizationPromise(
        "NativeAuthSessionError", IoLoginError.Type.MISSING_ACTIVITY_ON_PREPARE
      )
    }
  }

  @ReactMethod
  fun supportsInAppBrowser(promise: Promise) {
    val customTabBrowserPackageName = BrowserPackageHelper.getPackageNameToUse(reactApplicationContext)
    val supportsCustomTabs = !customTabBrowserPackageName.isNullOrEmpty()
    promise.resolve(supportsCustomTabs)
  }

  @ReactMethod
  @Throws(IOException::class)
  fun getRedirects(
    url: String,
    headers: ReadableMap,
    callbackURLParameter: String?,
    promise: Promise
  ) {

    val urlArray = arrayListOf<String>()

    try {
      findRedirects(url, headers, urlArray,  callbackURLParameter, promise) { result ->
        val urls = result.toTypedArray()
        val resultArray: WritableArray = Arguments.fromArray(urls)
        promise.resolve(resultArray)
      }
    } catch (e: IOException) {
      promise.reject(
        "NativeRedirectError",
        "See user info",
        generateErrorUserInfo(IoLoginError.Type.FIRST_REQUEST_ERROR)
      )
      return
    }
  }

  private fun findRedirects(
    url: String,
    headers: ReadableMap? = null,
    urlArray: ArrayList<String>,
    callbackURLParameter: String?,
    promise: Promise,
    onComplete: (urlArray: ArrayList<String>) -> Unit
  ) {
    try {
      val connection = URL(url).openConnection() as HttpURLConnection
      connection.instanceFollowRedirects = false
      headers?.toHashMap()?.forEach { (key, value) ->
        connection.setRequestProperty(key, value.toString())
      }
      handleRedirects(connection, url, urlArray, callbackURLParameter, promise, onComplete)
    } catch (e: Exception) {
      promise.reject(
        "NativeRedirectError",
        "See user info",
        generateErrorUserInfo(IoLoginError.Type.CONNECTION_REDIRECT_ERROR)
      )
      return
    }
  }

  //region redirects
  private fun findRedirects(
    url: String,
    urlArray: ArrayList<String>,
    callbackURLParameter: String?,
    promise: Promise,
    onComplete: (urlArray: ArrayList<String>) -> Unit
  ) {
    findRedirects(url, null, urlArray, callbackURLParameter, promise, onComplete)
  }

  private fun handleRedirects(
    connection: HttpURLConnection,
    url: String,
    urlArray: ArrayList<String>,
    callbackURLParameter: String?,
    promise: Promise,
    onComplete: (urlArray: ArrayList<String>) -> Unit
  ) {
    val responseCode = connection.responseCode
    val serverHeaders = connection.headerFields
    if (BuildConfig.DEBUG) {
      for ((key, values) in serverHeaders) {
        debugLog(">>> $key: $values")
      }
    }
    val setCookieHeader = serverHeaders["Set-Cookie"] ?: emptyList()

    val webkitCookieManager = android.webkit.CookieManager.getInstance()
    webkitCookieManager.setAcceptCookie(true)

    // Sync each cookie to Android WebView/WebKit
    syncCookies(url, setCookieHeader) {
      if (BuildConfig.DEBUG) {
        val lookForWebKitCookies = webkitCookieManager.getCookie(url)
        debugLog("<<< $lookForWebKitCookies")
      }

      if (responseCode in 300..399) {
        var redirectUrl = connection.getHeaderField("Location")
        if (redirectUrl.startsWith("/")) {
          val previousUrl = URL(url)
          val redirectScheme = previousUrl.protocol
          val redirectHost = previousUrl.host
          val port = previousUrl.port.toString()
          redirectUrl =
            redirectScheme + "://" + redirectHost + (if (port == "-1") "" else ":$port") + redirectUrl
        }
        urlArray.add(redirectUrl)
        if (getUrlParameter(redirectUrl).contains(callbackURLParameter)) {
          onComplete(urlArray)
          return@syncCookies
        } else {
          findRedirects(redirectUrl, urlArray, callbackURLParameter, promise, onComplete)
          return@syncCookies
        }
      } else if (responseCode >= 400) {
        promise.reject(
          "NativeRedirectError",
          "See user info",
          generateErrorUserInfo(
            IoLoginError.Type.REDIRECTING_ERROR,
            responseCode,
            url = getUrlWithoutQuery(url),
            parameters = getUrlParameter(url)
          )
        )
        return@syncCookies
      }
    }
  }

  private fun getUrlParameter(url: String): List<String> {
    val urlAsURL = URL(url)
    val parameters = mutableListOf<String>()
    urlAsURL.query?.let { queryString ->
      queryString.split("&")
        .forEach { queryParam ->
          val param = queryParam.split("=")
          parameters.add(param[0])
        }
    }
    return parameters
  }

  private fun getUrlWithoutQuery(url: String): String {
    val urlAsURL = URL(url)
    return "${urlAsURL.protocol}://${urlAsURL.authority}${urlAsURL.path}"
  }

  private fun debugLog(message: String) {
    if (BuildConfig.DEBUG) {
      println(message)
    }
  }

  private fun syncCookies(url: String, cookies: List<String>, onComplete: () -> Unit) {
    val webkitCookieManager = android.webkit.CookieManager.getInstance()
    webkitCookieManager.setAcceptCookie(true)

    fun setNext(index: Int) {
      if (index >= cookies.size) {
        webkitCookieManager.flush()
        onComplete()
        return
      }
      val cookieString = cookies[index]
      debugLog("$$$ Cookie string: $cookieString")
      webkitCookieManager.setCookie(url, cookieString) { _ /*success true/false*/ ->
        setNext(index + 1)
      }
    }
    setNext(0)
  }
  //endregion

  companion object {
    const val NAME = "IoLoginUtils"

    var authorizationPromise: Promise? = null

    fun rejectAndClearAuthorizationPromise(code: String, errorType: IoLoginError.Type) {
      authorizationPromise?.reject(
        code,
        "See user info",
        generateErrorUserInfo(errorType)
      )
      authorizationPromise = null
    }
  }

  override fun getName() = NAME
}
